Sentry ratelimit cannot be changed when using self-hosted

## Important Details

How are you running Sentry?

<!-- Please pick one of the following -->
On-Premise wo/ Docker, version 20.8.0

## Description
Sentry ratelimit cannot be changed when running on premises

## Steps to Reproduce

1. Go to web-interface, Admin/Settings
2. Set a non-zero ratelimit.
3. Get an error.

````
Oct  6 07:18:49 jsentry sentry[4128]: 10.100.33.5 - - [06/Oct/2020:04:18:49 +0000] "GET /api/0/internal/options/ HTTP/1.1" 200 20407 "https://sentry.findmykids.org/manage/settings/" "Mozilla/5.0 (X11; FreeBSD amd64; rv:76.0) Gecko/20100101 Firefox/76.0"
Oct  6 07:19:09 jsentry sentry[4128]: Traceback (most recent call last):
Oct  6 07:19:09 jsentry sentry[4128]:   File "/usr/local/lib/python2.7/site-packages/sentry-20.8.0-py2.7.egg/sentry/api/base.py", line 134, in handle_exception
Oct  6 07:19:09 jsentry sentry[4128]:     response = super(Endpoint, self).handle_exception(exc)
Oct  6 07:19:09 jsentry sentry[4128]:   File "/usr/local/lib/python2.7/site-packages/djangorestframework-3.6.4-py2.7.egg/rest_framework/views.py", line 449, in handle_exception
Oct  6 07:19:09 jsentry sentry[4128]:     self.raise_uncaught_exception(exc)
Oct  6 07:19:09 jsentry sentry[4128]:   File "/usr/local/lib/python2.7/site-packages/sentry-20.8.0-py2.7.egg/sentry/api/base.py", line 247, in dispatch
Oct  6 07:19:09 jsentry sentry[4128]:     response = handler(request, *args, **kwargs)
Oct  6 07:19:09 jsentry sentry[4128]:   File "/usr/local/lib/python2.7/site-packages/sentry-20.8.0-py2.7.egg/sentry/api/endpoints/system_options.py", line 74, in put
Oct  6 07:19:09 jsentry sentry[4128]:     options.set(k, v)
Oct  6 07:19:09 jsentry sentry[4128]:   File "/usr/local/lib/python2.7/site-packages/sentry-20.8.0-py2.7.egg/sentry/options/manager.py", line 83, in set
Oct  6 07:19:09 jsentry sentry[4128]:     "%r cannot be changed at runtime because it is configured on disk" % key
Oct  6 07:19:09 jsentry sentry[4128]: AssertionError: u'system.url-prefix' cannot be changed at runtime because it is configured on disk
Oct  6 07:19:09 jsentry sentry[4128]: 10.100.33.5 - - [06/Oct/2020:04:19:09 +0000] "PUT /api/0/internal/options/ HTTP/1.1" 500 746 "https://sentry.findmykids.org/manage/settings/" "Mozilla/5.0 (X11; FreeBSD amd64; rv:76.0) Gecko/20100101 Firefox/76.0"
````

### What you expected to happen
Ratelimit should be changeable from web-interface.

### Possible Solution
Stop sending system.url-prefix in PUT request ?

--- a/src/sentry/api/endpoints/system_options.py
+++ b/src/sentry/api/endpoints/system_options.py
@@ -72,10 +72,13 @@ class SystemOptionsEndpoint(Endpoint):
                     options.delete(k)
                 else:
                     options.set(k, v)
-            except TypeError as e:
+            except (TypeError, AssertionError) as e:
+                # TODO(chadwhitacre): Use a custom exception for the
+                # immutability case, especially since asserts disappear with
+                # `python -O`.
                 return Response(
                     {
-                        "error": "invalid_type",
+                        "error": "invalid_type" if type(e) is TypeError else "immutable_option",
                         "errorDetail": {"option": k, "message": six.text_type(e)},
                     },
                     status=400,

--- a/src/sentry/static/sentry/app/components/forms/apiForm.tsx
+++ b/src/sentry/static/sentry/app/components/forms/apiForm.tsx
@@ -1,3 +1,4 @@
+import React from 'react';
 import PropTypes from 'prop-types';
 
 import {
@@ -7,10 +8,12 @@ import {
 } from 'app/actionCreators/indicator';
 import {APIRequestMethod, Client} from 'app/api';
 import Form from 'app/components/forms/form';
+import FormField from 'app/components/forms/formField';
 import FormState from 'app/components/forms/state';
 import {t} from 'app/locale';
 
 type Props = Form['props'] & {
+  omitDisabled?: boolean;
   onSubmit?: (data: object) => void;
   apiEndpoint: string;
   apiMethod: APIRequestMethod;
@@ -23,6 +26,7 @@ export default class ApiForm extends Form<Props> {
 
   static propTypes = {
     ...Form.propTypes,
+    omitDisabled: PropTypes.bool,
     onSubmit: PropTypes.func,
     apiMethod: PropTypes.string.isRequired,
     apiEndpoint: PropTypes.string.isRequired,
@@ -32,12 +36,35 @@ export default class ApiForm extends Form<Props> {
 
   static defaultProps = {
     ...Form.defaultProps,
+    omitDisabled: false, // TODO(chadwhitacre) Upstream, flip to true, deprecate.
     submitErrorMessage: t('There was an error saving your changes.'),
     submitLoadingMessage: t('Saving changes\u2026'),
   };
 
   componentWillUnmount() {
     this.api.clear();
+  }
+
+  getEnabledData() {
+    // Return a hash of data from non-disabled fields.
+
+    // Start with this.state.data and remove rather than starting from scratch
+    // and adding, because a) this.state.data is our source of truth, and b)
+    // we'd have to do more work to loop over the state.data Object and lookup
+    // against the props.children Array (looping over the Array and looking up
+    // in the Object is more natural). Maybe the consequent use of delete
+    // carries a slight performance hit. Why is yer form so big? 🤔
+
+    const data = {...this.state.data}; // Copy to avoid mutating state.data itself.
+    React.Children.forEach(this.props.children, (child: any) => {
+      if (!FormField.isPrototypeOf(child.type)) {
+        return; // Form children include h4's, etc.
+      }
+      if (child.key && child.props?.disabled) {
+        delete data[child.key]; // Assume a link between child.key and data. 🐭
+      }
+    });
+    return data;
   }
 
   onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
@@ -47,7 +74,12 @@ export default class ApiForm extends Form<Props> {
       return;
     }
 
-    const {data} = this.state;
+    // Actual HTML forms do not submit data for disabled fields, and because of
+    // the way some of our APIs are implemented, we need to start doing the
+    // same. But, since some other parts of the app very probably depend on
+    // sending disabled fields, keep that the default for now.
+    // TODO(chadwhitacre): Expand and upstream this.
+    const data = this.props.omitDisabled ? this.getEnabledData() : this.state.data;
 
     this.props.onSubmit && this.props.onSubmit(data);
     this.setState(

--- a/src/sentry/static/sentry/app/components/forms/form.tsx
+++ b/src/sentry/static/sentry/app/components/forms/form.tsx
@@ -112,24 +112,10 @@ class Form<
   };
 
   onSubmitSuccess = (data: object) => {
-    const curData = this.state.data;
-    let newData = {};
-    if (data) {
-      Object.keys(curData).forEach(k => {
-        if (data.hasOwnProperty(k)) {
-          newData[k] = data[k];
-        } else {
-          newData[k] = curData[k];
-        }
-      });
-    } else {
-      newData = curData;
-    }
-
     this.setState({
       state: FormState.READY,
       errors: {},
-      initialData: newData,
+      initialData: {...this.state.data, ...(data || {})},
     });
     this.props.onSubmitSuccess && this.props.onSubmitSuccess(data);
   };

--- a/src/sentry/static/sentry/app/views/admin/adminSettings.tsx
+++ b/src/sentry/static/sentry/app/views/admin/adminSettings.tsx
@@ -66,6 +66,7 @@ export default class AdminSettings extends AsyncView<{}, State> {
           apiMethod="PUT"
           apiEndpoint={this.endpoint}
           initialData={initialData}
+          omitDisabled
           requireChanges
         >
           <h4>General</h4>

--- a/tests/sentry/api/endpoints/test_system_options.py
+++ b/tests/sentry/api/endpoints/test_system_options.py
@@ -67,6 +67,13 @@ class SystemOptionsTest(APITestCase):
         assert response.status_code == 400
         assert response.data["error"] == "unknown_option"
 
+    def test_put_hardwired_option(self):
+        with self.settings(SENTRY_OPTIONS={"system.url-prefix": "cheese"}):
+            self.login_as(user=self.user, superuser=True)
+            response = self.client.put(self.url, {"system.url-prefix": "bread"})
+            assert response.status_code == 400
+            assert response.data["error"] == "immutable_option"
+
     def test_put_simple(self):
         self.login_as(user=self.user, superuser=True)
         assert options.get("mail.host") != "lolcalhost"

